master/internal/trials/postgres_trials.go

```diff
@@ -138,3 +144,259 @@ func MetricsTimeSeries(trialID int32, startTime time.Time,
 	}
 	return metricMeasurements, nil
 }
+
+// Trial is a better bun trial model than the one in pkg/model/experiment.go.
+type Trial struct {
+	bun.BaseModel             `bun:"table:trials"`
+	ID                        int `bun:"id,pk,autoincrement"`
+	ExperimentID              int
+	State                     model.State
+	StartTime                 time.Time
+	EndTime                   *time.Time
+	Hparams                   map[string]any
+	WarmStartCheckpointID     *int
+	Seed                      int
+	RequestID                 *string
+	BestValidationID          *int
+	RunnerState               string // TODO(ilia): enum for training/validating/checkpointing.
+	TaskID                    *string
+	RunID                     int
+	Restarts                  int
+	Tags                      map[string]string // Note: values are always "".
+	CheckpointSize            int
+	CheckpointCount           int
+	SearcherMetricValue       *float64
+	SearcherMetricValueSigned *float64
+	TotalBatches              int
+	SummaryMetrics            map[string]any // TODO(ilia): better typing.
+	SummaryMetricsTimestamp   *time.Time
+	LatestValidationID        int
+	LastActivity              *time.Time
+	ExternalTrialID           *string
+}
+
+// LatestCheckpointForTrialTx finds the latest completed checkpoint for a trial, returning nil if
+// none exists.
+func LatestCheckpointForTrialTx(ctx context.Context, idb bun.IDB, trialID int) (
+	*model.Checkpoint, error,
+) {
+	var checkpoint model.Checkpoint
+	err := idb.NewSelect().Model(&checkpoint).
+		Where("trial_id = ?", trialID).
+		Where("state = 'COMPLETED'").
+		Order("steps_completed DESC").
+		Limit(1).
+		Scan(ctx)
+	if err != nil {
+		return nil, db.MatchSentinelError(err)
+	}
+	return &checkpoint, nil
+}
+
+// UpdateUnmanagedExperimentStatesTx updates an [unmanaged] experiment state according to its
+// constituent trial states.
+func UpdateUnmanagedExperimentStatesTx(
+	ctx context.Context, tx bun.IDB, experiments []*model.Experiment,
+) error {
+	var trialsRes []Trial
+	experimentIDs := make([]int, 0, len(experiments))
+	for _, exp := range experiments {
+		experimentIDs = append(experimentIDs, exp.ID)
+	}
+
+	if err := tx.NewSelect().Model(&trialsRes).
+		Column("id", "state", "end_time", "experiment_id").
+		Where("experiment_id in (?)", bun.In(experimentIDs)).
+		Order("end_time DESC NULLS LAST").
+		Scan(ctx); err != nil {
```

Comment:
very dicey whether this needs for update or not. trying to think about phantom reads if new trials suddenly appear in the `experiment_id in (?)` scan.

Reply:
i guess in theory, one of the two transactions would see all the rows that feed into the aggregate and then result would be correct.